### PR TITLE
Deprecate the `process_file_in_src` and `process_file` functions.

### PR DESCRIPTION
**Pull request by ltratt**
The `CTBuilder` API has always been a bit unsatisfactory, and some changes I'd like to make to it in the near future have made clear to me that we need some bigger changes.

https://github.com/softdevteam/grmtools/commit/4d6625c9b8210e9ac77aa6b905ec23e4bd12aa81 solves the problem that the `process_*` functions take arguments that really should have been options on the builder itself. In practise this means that `CTBuilder::process_file_in_src` is deprecated in favour of `CTBuilder::grammar_path_in_src` and `CTBuilder::process`. In most cases this means changing a `build.rs` from:

```rust
let lex_rule_ids_map = CTParserBuilder::new()
.process_file_in_src("grm.y")?;
```

to:

```rust
let lex_rule_ids_map = CTParserBuilder::new()
.grammar_path_in_src("grm.y")?
.process()?;
```

The less commonly used `process_file` function is similarly deprecated in favour of the `grammar_path`, `output_path`, and `process` functions. To my slight horror, some of the cttests used `process_file` in a way that only really worked by accident: by explicitly using `grammar_path` we lessen that chance. https://github.com/softdevteam/grmtools/commit/11981b6ee000eeb23a1d01b142b3ac397988254a ensures that.

While I'm here, I've tidied up some of the documentation that surrounds these functions: there's still more that could, and probably should, be done but that's best left to later commits.

This is a backwards compatible change in the sense that you can use the old functions but rustc will warn you about them. That should help people migrate their code more easily than a "hard" API change.

### CODE REVIEW

**Comment by ltratt** [comment]
bors try

**Comment by ltratt** [comment]
bors ping

**Comment by ltratt** [comment]
bors ping

**Comment by ltratt** [comment]
bors try

**Comment by ltratt** [comment]
That last commit (at least) will need squashing away.

bors try

**Comment by ltratt** [comment]
I'd like to pretend I was deliberately testing that https://github.com/ltratt/grmtools/commit/a125a70b2d3852ed1cf12d14608a9ddcc413b47b works, but https://github.com/ltratt/grmtools/commit/cb23764e5cd1cee1f11346e8f65f2df07c7fd67a is just the result of my idiocy.

bors try

**Comment by ltratt** [comment]
@ptersilie This is ready for review (but will definitely need squashing!).

File: CHANGES.md [hunk 1]
```diff
@@ -1,3 +1,20 @@
+# grmtools 0.10.3 (XXXX-YY-ZZ)
+
+* `CTBuilder::process_file_in_src` is deprecated in favour of
+  `CTBuilder::grammar_path_in_src` and `process`. In most cases you can change
```
Comment (ptersilie) [inline CHANGES.md]: I'm unsure about the name `grammar_path_in_src`. To me it suggest we give it a source which contains the grammar path. Maybe it should just be called `grammar_src`, since we are passing a grammar file, not a path.

Comment (ltratt) [reply CHANGES.md]: I'm also not sure what the best name is, and I was curious what you would think! I think `grammar_src` isn't good because that can be confused with `grammar_path`: I think we need to get the notion of "this file is relative to `src/`" across somehow (but without being too wordy). Maybe something like `grammar_file_in_src` or even `find_grammar_in_src`?

Comment (ptersilie) [reply CHANGES.md]: After reading the rest of the commit I realised that `src` was referencing the project's `src` directory. However, that is not always correct. What is actually used is the current working directory which isn't necessarily the project root. Maybe `grammar_file_in_cwd` would me more accurate, though it is a bit of a mouthful.

What does `grammar_path` take as an argument? I think you called it `path` because it takes a rust `Path`, but is that clear enough to the end user? When I read `path` I assume it takes a path and not for example a file.

File: lrpar/src/lib/ctbuilder.rs [hunk 2]
```diff
@@ -197,9 +197,9 @@ where
         }
     }
 
-    /// Set the input grammar path to a file in this project's `src` directory. This will also set
-    /// the output path (i.e. you do not need to call [CTParserBuilder::output_path]) and module
-    /// name (i.e. you do not need to call [CTParserBuilder::mod_name]).
+    /// Set the input grammar path to a file relative to this project's `src` directory. This will
```
Comment (ptersilie) [inline lrpar/src/lib/ctbuilder.rs:200]: This is what I was referring to in my earlier comment. The doc string suggests this is always relative to the project root, but that's not true if the binary is run from somewhere else, e.g. `/home/lukas/grmtools/bin/grmtools grm.y`.

File: CHANGES.md (hunk 1 above)
Comment (ltratt) [reply CHANGES.md]: > After reading the rest of the commit I realised that src was referencing the project's src directory. However, that is not always correct. What is actually used is the current working directory which isn't necessarily the project root.

Are there cases when the CWD isn't the project root? If so, is there a way to genuinely use the project root?

> What does grammar_path take as an argument?

`AsRef<Path>`. It's a common idiom in the Rust libraries, but a bit hard to describe succinctly, because it can auto-transform strings to `Path`s.

File: lrpar/src/lib/ctbuilder.rs (hunk 2 above)
Comment (ltratt) [reply lrpar/src/lib/ctbuilder.rs:200]: There is no `grmtools` binary so I'm not sure this can really happen.

File: CHANGES.md (hunk 1 above)
Comment (ptersilie) [reply CHANGES.md]: > `AsRef<Path>`. It's a common idiom in the Rust libraries, but a bit hard to describe succinctly, because it can auto-transform strings to `Path`s.

I meant, what does the user pass in, not the type, but the destination. Do they pass a file or really just a path. I think it's the former, that's why I pointed out that it might be confusing to the end user, that the function is called `grammar_path` when actually they are passing in a file.

File: lrpar/src/lib/ctbuilder.rs (hunk 2 above)
Comment (ptersilie) [reply lrpar/src/lib/ctbuilder.rs:200]: This was just an example. There will be some binary that the user runs, and since they can run it from anywhere, I feel like claiming this is always relative to the project root is misleading. Unless you mean it's only possible to run this from within the project's root due to other dependencies?

File: CHANGES.md (hunk 1 above)
Comment (ltratt) [reply CHANGES.md]: Technically they pass in a relative path.

File: lrpar/src/lib/ctbuilder.rs (hunk 2 above)
Comment (ltratt) [reply lrpar/src/lib/ctbuilder.rs:200]: I'm sure someone can work out how to run it in a weird way, but it's only documented as being runnable from a project's `build.rs` and I think there that CWD == project root (but I might be wrong).

Comment (ptersilie) [reply lrpar/src/lib/ctbuilder.rs:200]: Ah, it depends on `build.rs` which takes the grammar path and the compiles the parser. In that case I agree that it's unlikely someone will run it from anywhere but the project root.

File: CHANGES.md (hunk 1 above)
Comment (ptersilie) [reply CHANGES.md]: To a file or directory?

Comment (ltratt) [reply CHANGES.md]: Can I suggest we address the questions to the methods' doc strings? `CHANGES.md` only includes a summary, not the complete specification.

Comment (ptersilie) [reply CHANGES.md]: What questions? I didn't see any in the doc strings. Anyway, I was just trying to help find clearer names for those methods. If you don't want to change them, that's fine.

Comment (ltratt) [reply CHANGES.md]: I mean: if you find the doc strings for these functions (they're all in this PR, as they've been changed), you might find that it helps give you ideas for the names. e.g. https://github.com/softdevteam/grmtools/blob/9cae583549012f11fe534a6f5f0ba608279ef535/lrpar/src/lib/ctbuilder.rs#L200

Comment (ptersilie) [reply CHANGES.md]: The only ideas I got left is maybe `local_grammar_file`, `relative_grammar_file`. My mind is a bit blank at the moment.

Comment (ltratt) [reply CHANGES.md]: I think `grammar_in_src_dir` might be the best bet?

Comment (ptersilie) [reply CHANGES.md]: I can live with that.

Comment (ltratt) [reply CHANGES.md]: See if https://github.com/softdevteam/grmtools/pull/247/commits/da8d4490ba6869892d189e4bb0b29ed7bec9a0bd works for you.

**Comment by ptersilie** [comment]
Please squash.

**Comment by ltratt** [comment]
Squashed.
